%
% uaThesis example (for a thesis written in Portuguese)
%
% the complete list of options and commands can be found in uaThesis.sty
%

\documentclass[11pt,twoside,a4paper]{report}
\usepackage[DETI,newLogo]{uaThesis}

\def\ThesisYear{2017}

% optional packages
\usepackage[utf8]{inputenc}
\usepackage[english]{babel}
\usepackage{hyperref}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{xspace}% used by \sigla
\usepackage{cite}

% optional (comment to use default)s
%   depth of the table of contents
%     1 ... chapther and sections
%     2 ... chapters, sections, and subsections
%     3 ... chapters, sections, subsections, and subsubsections
\setcounter{tocdepth}{3}

% optional (comment to used default)
%   horizontal line to separate floats (figures and tables) from text
\def\topfigrule{\kern 7.8pt \hrule width\textwidth\kern -8.2pt\relax}
\def\dblfigrule{\kern 7.8pt \hrule width\textwidth\kern -8.2pt\relax}
\def\botfigrule{\kern -7.8pt \hrule width\textwidth\kern 8.2pt\relax}

% custom macros (could also be defined using \newcommand)
\def\I{\mathtt{i}}         % one possible way to represent $\sqrt{-1}$
\def\Exp#1{e^{2\pi\I #1}}  % argument inside braces, i.e., "{}"
\def\EXP#1.{e^{2\pi\I #1}} % argument finishes when a full stop is encountered, i.e., "."
\def\sigla{\LaTeX\xspace}  % use as "blabla \sigla blabla (no need to do "blabla \sigla\ blabla"

\def\AddVMargin#1{\setbox0=\hbox{#1}%
                  \dimen0=\ht0\advance\dimen0 by 2pt\ht0=\dimen0%
                  \dimen0=\dp0\advance\dimen0 by 2pt\dp0=\dimen0%
                  \box0}   % add extra vertical space above and below the argument (#1)
\def\Header#1#2{\setbox1=\hbox{#1}\setbox2=\hbox{#2}%
           \ifdim\wd1>\wd2\dimen0=\wd1\else\dimen0=\wd2\fi%
           \AddVMargin{\parbox{\dimen0}{\centering #1\\#2}}} % put #1 on top #2


\begin{document}

%
% Cover page (use only one of the first two \TitlePage)
%

% First alternative, with a figure
\TitlePage
  %\GRID  % for debugging ONLY
  \HEADER{\BAR\FIG{\includegraphics[height=60mm]{uaLogoNew}}} % the \FIG{} is optional
         {\ThesisYear}
  \TITLE{Rui Espinha\newline Ribeiro}
        {Multiple Sequence Alignment for Mitochondrial Orchestra}
\EndTitlePage
\titlepage\ \endtitlepage % empty page

% Second alternative, with a citation
\TitlePage
  %\GRID  % for debugging ONLY
  \HEADER{\BAR\FIG{\includegraphics[height=60mm]{uaLogoNew}}} % the \FIG{} is optional
         {\ThesisYear}
  \TITLE{Rui Espinha\newline Ribeiro}
        {Alinhamento  Sequencial M\'ultiplo para Orquestra Mitocondrial}
\EndTitlePage
\titlepage\ \endtitlepage % empty page


%
% Initial thesis pages
%

\TitlePage
  \HEADER{}{\ThesisYear}
  \TITLE{Rui Espinha\newline Ribeiro}
        {Multiple Sequence Alignment for Mitochondrial Orchestra}
  \vspace*{15mm}
  \TEXT{}
       {Disserta\c c\~ao apresentada \`a Universidade de Aveiro para cumprimento dos requesitos
        necess\'arios \`a obten\c c\~ao do grau de Mestre em Engenharia de Computadores e Telem\'atica, realizada sob a orienta\c c\~ao
        cient\'\i fica do Doutor Carlos Alberto da Costa Bastos, Professor do Departamento de Eletr\'onica, Telecomunica\c c\~oes e Inform\'atica da Universidade de Aveiro e da Doutora Vera M\'onica Almeida Afreixo, Professora do Departamento de Matem\'atica da Universidade de Aveiro.}
\EndTitlePage
\titlepage\ \endtitlepage % empty page

\TitlePage
  \vspace*{55mm}
  \TEXT{\textbf{o j\'uri~/~the jury\newline}}
       {}
  \TEXT{presidente~/~president}
       {\textbf{ABC}\newline {\small
        Professor Catedr\'atico da Universidade de Aveiro (por delega\c c\~ao da Reitora da
        Universidade de Aveiro)}}
  \vspace*{5mm}
  \TEXT{vogais~/~examiners committee}
       {\textbf{DEF}\newline {\small
        Professor Catedr\'atico da Universidade de Aveiro (orientador)}}
  \vspace*{5mm}
  \TEXT{}
       {\textbf{GHI}\newline {\small
        Professor associado da Universidade J (co-orientador)}}
  \vspace*{5mm}
  \TEXT{}
       {\textbf{KLM}\newline {\small
        Professor Catedr\'atico da Universidade N}}
\EndTitlePage
\titlepage\ \endtitlepage % empty page

\TitlePage
  \vspace*{55mm}
  \TEXT{\textbf{agradecimentos~/\newline acknowledgements}}
       {NO ONE AT ALL}
\EndTitlePage
\titlepage\ \endtitlepage % empty page

\TitlePage
  \vspace*{55mm}
  \TEXT{\textbf{Resumo}}
       {Um bom resumo da minha tese.}
\EndTitlePage
\titlepage\ \endtitlepage % empty page

\TitlePage
  \vspace*{55mm}
  \TEXT{\textbf{Abstract}}
       {An abstract for the ages}
\EndTitlePage

%\EndTitlePage
\titlepage\ \endtitlepage % empty page


%
% Tables of contents, of figures, ...
%

\pagenumbering{roman}
\tableofcontents

\cleardoublepage
\listoffigures

\cleardoublepage
\listoftables


% The chapters (usually written using the isolatin font encoding ...)

\cleardoublepage
\pagenumbering{arabic}
\chapter{Introduction}
Over the last two decades, there has been a growing focus on an atipical way of analysing genetic information to establish any relation between two species, the so-called "genomic music". Although we can speculate about the scientific and artistic interest of such studies, only the first one is actually explored in this work's scope as the second one and its subsequent discussions are a consequence of the musical elements we introduce to the game. 
\\\\The proposal for this thesis is to build a bridge between phylogenetic studies and music through data science algorithms, as the main intent is to explore associations between different genetic codes which enable us to assume a relation between two different species. In practice, the final work is an application that creates a virtual orchestra with the mitochondrial genetic code. The instruments and the music's dynamics should be binded with the subspecies found in a specific genetic code, allowing us to give a specie a certain musical identity.\\ 
\\This allows to associate separate species by simply listening them, which enables any non-science related individual to take some sort of conclusion by listening to "the sound of a specie". It can also serve as a contribution to the long-term scientific battle that is the study of data science techniques that find evidence of relations in different genetic codes, which obviously can bring benefits in future research works.


\section{Motivation}
As Clare Sansom states in \textit{The Biochemist}, "Parts of DNA and protein sequences are often repeated with only small changes, and this imperfect repetition has echoes in the themes and variations found in classical music.".
E mais coisas que ainda nao escrevi....

\section{Genomic music}
\subsection{Scientific background}
\subsection{Existing projects}
\subsection{Music from the genetic code: what does it mean in 2016?}

\section{Interdisciplinary concepts: Biology}
\subsection{Basic notions}
\subsection{Multiple Sequence Alignment}
One of the major tasks in this work is to find homologous sequences so we can use them to map music in a coherent way. By homologous, we mean as having structural and evolutionary resemblance - a shared ancestry between them. \cite[chapter, p.~215]{biodurbin} Such techniques for estimating homologous regions are called alignments. More specifically, the global alignment of such two sequences is called a pairwise alignment. But in this work we intent to find homologous regions between large sets of mitochondrial DNA sequences. In this case, our goal is to perform multiple sequence alignment methods (MSA, as we will refer to them in descriptions, from this chapter on).\\
The analysis of these methods is highly complex as their correctness depends on the relatedness of our set of sequences. Alignment techniques must always keep in mind two key features: the fact that some positions (in the alignment) are more conserved than others, e.g. position-specific scoring; and the fact that the sequences are not independent, but instead are related by a phylogenetic tree (which we will explore more ahead) (R. Durbin et al.). MSA algorithms usually include indels (or ``gaps'', as they are commonly known). Such gaps are relative to insertion or deletion events that are represented between characters in an alignment column.\\
There are 4 main types of MSA methods:
\begin{enumerate}
\item Dynamic programming: this is the most direct approach to MSA. As the name suggests, it uses dynamic programming techniques (mainly, storing computationally complex results in a data structure): we usually have a gap penalty and a substitution matrix (the storage structure). The score (probability) of an aligment between each pair of amino acids are mapped in the matrix based on the similarity of their chemical properties and evolutionary probability of the mutation. This method is the result of the generalisation of the dynamic programming alignment approach that exists for pairwise alignments.\\
This process can have several performance issues: it demans the computation of the whole substitution matrix, which implies that for a set of N sequences, we always estimate $2^{n}-1$ pairwise alignments in a $L.N$ matrix, where L is the length of the sequences. 

[A PICTURE THAT SHOWS WHAT'S HAPPENING]
\item Progressive Alignment
\item Iterative Methods
\item Hidden Markov Models (HMM)
\end{enumerate}

\begin{description}
\item Furthermore: Consensus Methods
\end{description}

\subsection{Phylogenetics}
\subsubsection{Distance matrices}
\subsubsection{Parsimony}
\subsubsection{Bayesian inference??}

\section{Interdisciplinary concepts: Music}
\subsection{Basic notions}
\subsection{Dynamics}
\subsection{Families of instruments in the orchestra}


\section{General Dissertation Structure}


\chapter{From code to music}
\section{Musical synthesis}
\section{Classifying music: Zipf's law}
\section{Mapping code into music}
\subsection{Frequency of occurrence}
\subsection{Distances between words}
\subsection{Dynamics}

\chapter{The virtual orchestra}
\section{Multiple Sequence Alignment}
\section{Assigning species to instruments}
\section{Building the orchestra}
\section{Data relations: what species exist in this orchestra?}

\chapter{Programming environment}
\section{Python as a data science language}
\section{BioPython}
\section{Data analysis methods}

\chapter{Final Application}
\section{Interface}
\section{Tests}
\subsection{Accuracy in building phylogenetic tree}
\subsection{Comparison between different species}
\subsection{Comparison between music mappings}
\subsection{Auditive comparison}


\chapter{Results}
\chapter{Conclusions}

%
% The bibliography
%
\cleardoublepage
%\iffalse
  % Use this is the final version
  %  unsrt produces numbered entries, sorted by order of citation
  %  plain produces numbered entries, sorted alphabetically
  %  other styles are possible (I recommend the harvard package)
  %\bibliographystyle{unsrt}
  \bibliographystyle{plain}
  \bibliography{tese.bib}% replace by the name of name of your .bib file
%\else
  % An example (the contents of the .bbl file)
%  \begin{thebibliography}{10}

%  \bibitem{Eliahou-1-1993-CLBNCL}
%  Shalom Eliahou.
%  \newblock The $3x+1$ problem: New lower bounds on nontrivial cycle lengths.
%  \newblock {\em Discrete Mathematics}, 118(1--3):45--56, 1993.

%  \bibitem{Garner-1981-1-OCA}
%  Lynn~E. Garner.
%  \newblock On the collatz $3n+1$ algorithm.
%  \newblock {\em Proceedings of the American Mathematical Society}, 82(1):19--22,
%    May 1981.
%  \end{thebibliography}
%\fi
\cleardoublepage


\end{document}
